OpenRA.Mods.Common/Traits/World/Pathfinder/PathCacheStorage.cs

```diff
@@ -0,0 +1,67 @@
+﻿using System.Collections.Generic;
+using System.Linq;
+
+namespace OpenRA.Mods.Common.Traits
+{
+	public class PathCacheStorage : ICacheStorage<List<CPos>>
+	{
+		class CachedPath
+		{
+			public List<CPos> Result;
+			public int Tick;
+		}
+
+		private const int MaxPathAge = 50;
+		private readonly IWorld world;
+		Dictionary<string, CachedPath> cachedPaths = new Dictionary<string, CachedPath>(1000);
+
+		public PathCacheStorage(IWorld world)
+		{
+			this.world = world;
+		}
+
+		public void Remove(string key)
+		{
+			cachedPaths.Remove(key);
+		}
+
+		public void Store(string key, List<CPos> data)
+		{
+			// Eventually clean up the cachedPaths dictionary
+			if (cachedPaths.Count >= 1000)
+			{
+				foreach (var cachedPath in cachedPaths.Where(p => IsExpired(p.Value)))
+					cachedPaths.Remove(cachedPath.Key);
```

Comment:
I got an exception during testing here (maybe you need a test caches at least 1000 paths - might also be worth allowing the cache size to be set in the ctor come to think of it).

Anyway, you're modifying the collection whilst enumerating it. You should create a copy by calling ToList to avoid that.